core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala

```diff
@@ -195,6 +195,191 @@ class TaskSchedulerImplSuite extends SparkFunSuite with LocalSparkContext with B
+      Seq(TaskLocation("host1", "exec1"))
+    )
+    taskScheduler.resourceOffers(IndexedSeq(WorkerOffer("exec1", "host1", 1)))
+    taskScheduler.submitTasks(taskSet)
+    taskScheduler
+  }
+
+  test("SPARK-18886 -  partial offers (isAllFreeResources = false) reset timer before " +
+    "any resources have been rejected") {
+    val clock = new ManualClock()
+    val taskScheduler = setupTaskScheduler(clock)
+    val advanceAmount = 2000
+
+    // by default, new partial resource (isAllFreeResources = false) offers reset timer
+    // if the resource is accepted
+    clock.advance(advanceAmount)
+    assert(taskScheduler
+      .resourceOffers(
+        IndexedSeq(WorkerOffer("exec1", "host1", 1)),
+        isAllFreeResources = false)
+      .flatten.length === 1)
+
+    // would advance to NODE_LOCAL locality if timer wasn't reset above
+    // Verifying this node local task is not accepted
+    clock.advance(advanceAmount)
+    assert(taskScheduler
+      .resourceOffers(
+        IndexedSeq(WorkerOffer("exec2", "host1", 1)))
+      .flatten.isEmpty)
+  }
+
+  test("SPARK-18886 -  delay scheduling timer is reset when it accepts all resources offered when" +
+    "isAllFreeResources = true") {
+    val clock = new ManualClock()
+    val taskScheduler = setupTaskScheduler(clock)
+    val advanceAmount = 2000
+
+    // timer is reset when tsm accepts all free resources offered to it
+    clock.advance(advanceAmount)
+    assert(taskScheduler
+      .resourceOffers(
+        IndexedSeq(WorkerOffer("exec1", "host1", 1)),
+        isAllFreeResources = true)
+      .flatten.length === 1)
+
+    // would advance to NODE_LOCAL locality if timer wasn't reset above
+    // Verifying this node local task is not accepted
+    clock.advance(advanceAmount)
+    assert(taskScheduler
+      .resourceOffers(
+        IndexedSeq(WorkerOffer("exec2", "host1", 1)))
+      .flatten.isEmpty)
+  }
+
+  test("SPARK-18886 - partial resource offers (isAllFreeResources = false) reset " +
+    "time if last full resource offer (isAllResources = true) was accepted as well as any " +
+    "following partial resource offers") {
+    val clock = new ManualClock()
+    val taskScheduler = setupTaskScheduler(clock)
+    val advanceAmount = 2000
+
+    assert(taskScheduler
+      .resourceOffers(
+        IndexedSeq(WorkerOffer("exec1", "host1", 1)),
+        isAllFreeResources = true)
+      .flatten.length === 1)
+
+    clock.advance(advanceAmount)
+    assert(taskScheduler
+      .resourceOffers(
+        IndexedSeq(WorkerOffer("exec1", "host1", 1)),
+        isAllFreeResources = false)
+      .flatten.length === 1)
+
+    clock.advance(advanceAmount)
+    assert(taskScheduler
+      .resourceOffers(
+        IndexedSeq(WorkerOffer("exec1", "host1", 1)),
+        isAllFreeResources = false)
+      .flatten.length === 1)
+
+    clock.advance(advanceAmount)
+    assert(taskScheduler
+      .resourceOffers(
+        IndexedSeq(WorkerOffer("exec2", "host1", 1)))
```

Comment:
put in explicit isAllFreeResources variable

Reply:
👍